### PR TITLE
ENH: vendor SerializableLock from dask and use as default backend lock, adapt tests

### LINKED ISSUE #8442: Cannot pickle '_thread.lock' object exception after DataArray transpose and copy operations from netCDF file.
### What is your issue?

I hit this issue while using [rioxarray](https://github.com/corteva/rioxarray) with a series of operations similar to those noted in this issue https://github.com/corteva/rioxarray/issues/614.  After looking through the `rioxarray` codebase a bit I was able to reproduce the issue with pure `xarray` operations.

If the Dataset is opened with the default `lock=True` settings,  [transposing](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.transpose.html) a DataArray's coordinates and then [copying](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.copy.html) the DataArray results in a `cannot pickle '_thread.lock' object` exception.

If the Dataset is opened with `lock=False`, no error is thrown.

This [sample notebook](https://nbviewer.org/gist/sharkinsspatial/c4512d2b8e53b6c42cd16c88f76e7325) reproduces the error.

This might be user error on my part, but it would be great to have some clarification on why `lock=False` is necessary here as my understanding was that this should only be necessary when using parallel write operations.

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] Closes #8442
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/backends/dask_lock.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# The code in this module is taken from dask. For reference,
```
Comment (jhamman) [inline xarray/backends/dask_lock.py:1]: can we put this in `locks.py` or, do something similar to a one of our compat module (`pdcompat.py`)?

Comment (kmuehlbauer) [reply xarray/backends/dask_lock.py]: I've moved the code into locks.py as suggested by @jhamman, thanks! Left a note and a permalink to the dask source.

File: doc/whats-new.rst [hunk 2]
```diff
@@ -44,6 +44,8 @@ Bug fixes
 
 - Reverse index output of bottleneck's rolling move_argmax/move_argmin functions (:issue:`8541`, :pull:`8552`).
   By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
+- Fork `SerializableLock` from dask and use as default lock for netcdf4 backends (:issue:`8442`, :pull:`8571`).
```
Comment (dcherian) [inline doc/whats-new.rst:47]:
```suggestion
- Vendor `SerializableLock` from dask and use as default lock for netcdf4 backends (:issue:`8442`, :pull:`8571`).
```

**Comment by dcherian** [comment]
Adding this lock might fix some of our xfails in `test_backends.py`. Mind taking a look at the windows ones?

**Comment by kmuehlbauer** [comment]
@dcherian Yes, can have a look the coming days. I'll test first a bit on my windows box to not have to debug on CI.

**Comment by dcherian** [comment]
In that case, let's merge and remove the xfails later!

**Pull request merged**